src/kde/kde.t.hh

```diff
@@ -0,0 +1,356 @@
+        const double h = band_width[i][0];
+        for (int j = 0; j<global_lower_bound; j++){
+            const double x = global_x_position[j];
+            const double u = (x0-x)/h;
+            normal[i] += (epan_kernel(u))/h;
+        }
+    }
+    // fetch upper ranks nonlocal contribution
+    for (int i = 0; i<local_size; i++){
+        const double x0 = position[i][0];
+        const double h = band_width[i][0];
+        for (int j = global_upper_bound; j<size; j++){
+            const double x = global_x_position[j];
+            const double u = (x0-x)/h;
+            normal[i] += (epan_kernel(u))/h;
+        }
+    }
+    
+    double reconstruction_conservation = 0.0;
+    // now apply the kernel to the local ranks
+    for (int i = 0; i<local_size; i++){
+        const double x0 = position[i][0];
+        const double h = band_width[i][0];
+        // fetch local contribution
+        for (int j = 0; j<local_size; j++){
+            const double x = position[j][0];
+            const double u = (x0-x)/h;
+            const double val = distribution[j]*(epan_kernel(u))/h/normal[i];
+            result[i] += val;
+            reconstruction_conservation += val;
+        }
+    }
+    // apply contribution to lower ranks
+    for (int i = 0; i<local_size; i++){
+        const double x0 = position[i][0];
+        const double h = band_width[i][0];
+        for (int j = 0; j<global_lower_bound; j++){
+            const double x = global_x_position[j];
+            const double u = (x0-x)/h;
+            const double val = global_distribution[j]*(epan_kernel(u))/h/normal[i];
+            result[i] += val;
+            reconstruction_conservation += val;
+        }
+    }
+    // apply contribution to upper ranks
+    for (int i = 0; i<local_size; i++){
+        const double x0 = position[i][0];
+        const double h = band_width[i][0];
+        for (int j = global_upper_bound; j<size; j++){
+            const double x = global_x_position[j];
+            const double u = (x0-x)/h;
+            const double val = global_distribution[j]*(epan_kernel(u))/h/normal[i];
+            result[i] += val;
+            reconstruction_conservation += val;
+        }
+    }
+
+    if(domain_decomposed){
+        // accumulate global contribution
+        rtt_c4::global_sum(reconstruction_conservation);
+    }
+
+    if(!rtt_dsxx::soft_equiv(reconstruction_conservation,0.0) and !rtt_dsxx::soft_equiv(global_conservation,0.0)){
+        // Totals are non-zero so scale the result for conservation
+        for (int i=0; i<local_size; i++)
+            result[i] *= global_conservation/reconstruction_conservation;
+    } else {
+        // a zero distribution is possible. If it occurs fall back to residual conservation;
+        const double res = global_conservation-reconstruction_conservation;
+        for (int i=0; i<local_size; i++)
+            result[i] += res/double(size);
+    }
+    
+    return result;
+    
```

Comment:
* Not seeing any DbC.  It is worth taking a pass through this code thinking about pre- and post- conditions.
* It might be good to see those documented in the comment block as well (`\pre`, `\post`).

Reply:
Done.